In [2]:
import pandas as pd
import numpy as np
import re

In [3]:
df = pd.read_csv("autos.csv",encoding='Latin-1')

In [4]:
df.head(15)

dateCrawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6   2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8   2016-03-22 16:51:34                                         Seat_Arosa   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10  2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11  2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   

    seller offerType   price   abtest vehicleType  yearOfRegistration  \
0   privat   Angebot  $5,000  control         bus                2004   
1   privat   Angebot  $8,500  control   limousine                1997   
2   privat   Angebot  $8,990     test   limousine                2009   
3   privat   Angebot  $4,350  control  kleinwagen                2007   
4   privat   Angebot  $1,350     test       kombi                2003   
5   privat   Angebot  $7,900     test         bus                2006   
6   privat   Angebot    $300     test   limousine                1995   
7   privat   Angebot  $1,990  control   limousine                1998   
8   privat   Angebot    $250     test         NaN                2000   
9   privat   Angebot    $590  control         bus                1997   
10  privat   Angebot    $999     test         NaN                2017   
11  privat   Angebot    $350  control         NaN                2000   
12  privat   Angebot  $5,299  control  kleinwagen                2010   
13  privat   Angebot  $1,350  control   limousine                1999   
14  privat   Angebot  $3,999     test  kleinwagen                2007   

      gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0     manuell      158   andere  150,000km                    3      lpg   
1   automatik      286      7er  150,000km                    6   benzin   
2     manuell      102     golf   70,000km                    7   benzin   
3   automatik       71   fortwo   70,000km                    6   benzin   
4     manuell        0    focus  150,000km                    7   benzin   
5   automatik      150  voyager  150,000km                    4   diesel   
6     manuell       90     golf  150,000km                    8   benzin   
7     manuell       90     golf  150,000km                   12   diesel   
8     manuell        0    arosa  150,000km                   10      NaN   
9     manuell       90   megane  150,000km                    7   benzin   
10    manuell       90      NaN  150,000km                    4   benzin   
11        NaN        0      NaN  150,000km                    0   benzin   
12  automatik       71   fortwo   50,000km                    9   benzin   
13    manuell      101       a3  150,000km                   11   benzin   
14    manuell       75     clio  150,000km                    9   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4      

In [5]:
df.describe(include="all")

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44905        50000.000000    47320  50000.000000  47242   
unique           8                 NaN        2           NaN    245   
top      limousine                 NaN  manuell           NaN   golf   
freq         12859                 NaN    36993           NaN   4024   
mean           NaN         2005.073280      NaN    116.355920    NaN   
std            NaN          105.712813      NaN    209.216627    NaN   
min            NaN         1000.000000      NaN      0.000000    NaN   
25%            NaN         1999.000000      NaN     70.000000    NaN   
50%            NaN         2003.000000      NaN    105.000000    NaN   
75%            NaN         2008.000000      NaN    150.000000    NaN   
max            NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  monthOfRegistration fuelType       brand notRepairedDamage  \
count       50000         50000.000000    45518       50000             40171   
unique         13                  NaN        7          40                 2   
top     150,000km                  NaN   benzin  volkswagen              nein   
freq        32424                  NaN    30107       10687             35232   
mean          NaN             5.723360      NaN         NaN               NaN   
std           NaN             3.711984      NaN         NaN               NaN   
min           NaN             0.000000      NaN         NaN               NaN   
25%           NaN             3.000000      NaN         NaN               NaN   
50%           NaN             6.000000      NaN         NaN               NaN   
75%           NaN             9.000000      NaN         NaN               NaN   
max           NaN            12.000000      NaN         NaN               NaN   

                dateCreated  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

Firstly we will clean the column names, make them uniform and in snake case for easier understanding and retrieval

In [6]:
new_cols = {'dateCrawled':'date_crawled','offerType':'offer_type',
           'vehicleType':'vehicle_type','yearOfRegistration':'registration_year',
           'powerPS':'power_PS','monthOfRegistration':'registration_month',
           'fuelType':'fuel_type','notRepairedDamage':'repaired_damage',
           'dateCreated':'date_created','nrOfPictures':'no_pictures',
           'postalCode':'postal_code','lastSeen':'last_seen'}
        
df.rename(columns=new_cols,inplace=True)
df.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'repaired_damage', 'date_created', 'no_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

The price and odometer columns need cleaning to make them of type numeric and emove '$' ',' 'kms' from the strings.

In [7]:
df.price = df.price.str.replace('$','')
df.price = df.price.str.replace(',','').astype('int32')
df.odometer = df.odometer.str.replace('km','')
df.odometer = df.odometer.str.replace(',','').astype('int32')
df.rename(columns={'odometer':'odometer_km'},inplace=True)

Looking at the disctribution of prices, the min - 0 and max - 9999999 seem as outliers. 0 being a cost is highly unlikely and could be those advertisements whose price point could not be retrieved. The 9999999 is on the extreme end and an absurd value.
The registration month takes values from 1 to 12. month value with 0s are outliers in this data, and such data will be pruned.
Similarly, registration year has two outlier values 1800 and 9999, these years arnt possible, thus pruning them.

In [8]:
df = df[df.price.between(1,12345678)]
df = df[df.registration_month != 0]
df = df[df.registration_year.between(1927,2019)]

All date/time columns are given in strings, this makes it difficult for information retrieval from such columns. Converting them to datetime data type

In [9]:
df.date_created = pd.to_datetime(df.date_created)
df.date_crawled = pd.to_datetime(df.date_crawled)
df.last_seen = pd.to_datetime(df.last_seen)

In [10]:
brands = df.brand.value_counts(normalize=True)[0:6].index

In [11]:
brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [12]:
mean_price_data = {}
for i in brands:
    p = df[df['brand'] == i].price.mean()
    mean_price_data[i]=p

In [13]:
mean_mileage_data = {}
for i in brands:
    p = df[df.brand == i].odometer_km.mean()
    mean_mileage_data[i] = p

In [14]:
mean_price = pd.Series(mean_price_data)
mean_mileage = pd.Series(mean_mileage_data)

In [15]:
df2 = pd.DataFrame(mean_price,columns=["mean_price"])
df2['mean_mileage'] = mean_mileage
df2

mean_price   mean_mileage
volkswagen     5904.897294  128360.930736
bmw            8837.038194  132584.763072
opel           5813.933333  128880.952381
mercedes_benz  8826.130854  130987.144169
audi           9680.716251  128940.809969
ford           7929.724274  123985.817942